rfcs/convergence/context-values.md

```diff
@@ -0,0 +1,227 @@
+
+_There is no recomended option yet._
+
+### Option 1: Make `render` functions hooks
+
+We can transform existing render functions to be React hooks:
+
+```diff
+-function renderBaz(state) {
++function useRenderBaz(state) {
+  const { open, inline } = state;
+  // ✅ now we can use hooks
+  const contextValue = React.useMemo(() => ({ inline, open }), [inline, open]);
+
+  // ...
+}
+```
+
+It's one of the easiest options, but it could cause issues in future as it breaks separation of concerns. For example, other hooks can be also called now in `useRender`:
+
+```tsx
+function useRenderBaz(state) {
+  const { open, inline } = state;
+
+  // 💣 other hooks can be also declared there
+  React.useEffect(() => {
+    // ...
+  }, [open]);
+}
+```
+
+- 👎 breaks separation of concerns as allows to use other hooks in render functions
+
+### Option 2: Use `useContextSelector()` everywhere
+
+For `context-selector` memoization is not required as it will happen in consumers. In this case we don't need to transform `render` functions to React hooks.
+
+```tsx
+function renderBaz(state) {
+  const { open, inline } = state;
+
+  return (
+    <SampleContext.Provider value={{ inline, open }}>
+      <div />
+    </SampleContext.Provider>
+  );
+}
+
+// ---
+
+function useBazItem() {
+  // ✅ only if "open"/"inline" will change a rerender will be triggered
+  const open = useContextSelector(SampleContext, value => value.open);
+  const inline = useContextSelector(SampleContext, value => value.inline);
+}
+```
+
+- 👍 Allows to keep existing API design
+
+- 👎 Even if it's enough performant option, it will be anyway slower than `React.useContext()` as we are evaluating additional code during render cycles
+- 👎 (_potentially_) This implementation is tested with React's Concurrent mode, but still could cause issues with React 18
+- 👎 Sooner or later `useContextSelector` will be implemented in React's core ([facebook/react#20646](https://github.com/facebook/react/pull/20646)) and it could have different API than our implementation
+
+### Option 3: Modify render functions to accept context value
+
+Another option is to modify a signature of render functions to include context values as the second param and compute them inside component.
+
+```diff
+function FooComponent() {
+  const state = useFooState()
+
++ const barContextValue = React.useMemo(/* ... */)
++ const bazContextValue = React.useMemo(/* ... */)
+
+- return renderFoo(state)
++ return renderFoo(state, { bar: barContextValue, baz: bazContextValue })
+}
+
+// ---
+
+-function renderFoo(state) {
+-  return <SampleContext.Provider value={state.bazContextValue} />
++function renderFoo(state, contextValues) {
++  return <SampleContext.Provider value={contextValues.baz} />
+}
+```
+
+To improve experience for customers and ourselves we can extract creation of values in a separate hook:
+
+```tsx
+function useFooContextValues() {
+  const barContextValue = React.useMemo(/* ... */);
```

Comment:
in general this kind of memoizing patterns can be solved completely outside reacts context so we don't need any hooks.

```
function memoizedContextSelector(value:T){....}

function FooComponent() {
  const state = useFooState();
  const contextValues = memoizedContextSelector(state);
...
}
```

Reply:
- memoization via `React.useMemo()` is context aware
- memoized results will be also cleaned up sooner or later

---

For example, `memoize-one` ([CodeSandbox](https://codesandbox.io/s/silent-hill-qfopw?file=/src/App.js)) will not work properly in this case and will compute values again:

![image](https://user-images.githubusercontent.com/14183168/128686861-34c1f765-dce2-4370-b605-9bc7a3dd4b65.png)

---

For example, `fast-memoize` will do exactly what we want ([CodeSandbox](https://codesandbox.io/s/nifty-visvesvaraya-fm9kl?file=/src/App.js)):

![image](https://user-images.githubusercontent.com/14183168/128687104-c0cd2803-96cb-4e1e-bae0-f47203dbde58.png)

**But** then you will need to find a way to how to clear cache as it will grow:

![image](https://user-images.githubusercontent.com/14183168/128687533-47452179-675a-4a16-a65a-aa5490aa016b.png)

A limit for entries will not work as you don't know how many components customer will decide to render.

**Then** we will custom cache/serializer for DOM & React elements as default one is based on `JSON.stringify()` that will give inaccurate results:

![image](https://user-images.githubusercontent.com/14183168/128687891-5329ae1b-e651-4434-b2fc-47714d637c9e.png)
![image](https://user-images.githubusercontent.com/14183168/128688017-2f778e3a-7991-4f47-a477-0c92a0fcbae9.png)

And **then** problem with clearing cache becomes a big headache as cache may contain detached HTML elements that will immediately cause memory leaks.